DEHPCommon/UserPreferenceHandler/UserPreferenceService/UserPreferenceService.cs

```diff
@@ -0,0 +1,176 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="UserPreferenceService.cs"company="RHEA System S.A.">
+//    Copyright(c) 2020 RHEA System S.A.
+//    Author: Sam Gerené, Alex Vorobiev, Alexander van Delft, Kamil Wojnowski, Nathanael Smiechowski.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+namespace DEHPCommon.UserPreferenceHandler.UserPreferenceService
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Reflection;
+
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Serialization;
+
+    using NLog;
+
+    public class UserPreferenceService : IUserPreferenceService
+    {
+        /// <summary>
+        /// The logger for the current class
+        /// </summary>
+        private static Logger logger = LogManager.GetCurrentClassLogger();
+
+        /// <summary>
+        /// Path to special windows "AppData" folder 
+        /// </summary>
+        public static string AppDataFolder = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
+
+        /// <summary>
+        /// Application configuration folder path.
+        /// </summary>
+        public static string ConfigurationDirectoryFolder = "UserPreferenceService";
+
+        /// <summary>
+        /// The setting file extension
+        /// </summary>
+        public const string SETTING_FILE_EXTENSION = ".settings.json";
+
+        /// <summary>
+        /// A dictionary used to store the user preference setting,
+        /// </summary>
+        private readonly Dictionary<string, UserPreference> userPreferenceSettings;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="UserPreferenceService"/>
+        /// </summary>
+        public UserPreferenceService()
+        {
+            this.userPreferenceSettings = new Dictionary<string, UserPreference>();
+        }
+
+        /// <summary>
+        /// Reads the <see cref="T"/> user preference in settings
+        /// </summary>
+        /// <typeparam name="T">A type of <see cref="UserPreference"/></typeparam>
+        /// <returns>
+        /// An instance of <see cref="UserPreference"/>
+        /// </returns>
+        public T Read<T>() where T : UserPreference
+        {
+            var assemblyName = this.QueryAssemblyTitle(typeof(T));
+
+            if (this.userPreferenceSettings.TryGetValue(assemblyName, out var result))
+            {
+                return result as T;
+            }
+
+            this.CheckConfigurationDirectory();
+
+            var path = Path.Combine(this.ApplicationConfigurationDirectory, assemblyName);
```

Comment:
not appdata please, as discussed into the same folder as where the app is, or a relative folder under it.